docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,76 @@
+
+## Motivation
+
+At present, the handling of VPN connectivity lacks standardization and consistency. There are instances when the VPN goes into No Signal due to a variety of connectivity-related issues, such as the network connection failing or the server location becoming unavailable. In these situations, users are often left without a clear explanation of what went wrong or guidance on how to troubleshoot and resolve the issue. They may encounter a background error message or lose signal without further information.
+
+The objective with the Connection Manager is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
+
+To avoid further complicating the already complex logic of the controller, this document proposes encapsulating this new functionality within the Connection Manager. This object will interact with both the controller and connection health, serving as a dedicated entity for managing and improving VPN connectivity. This approach ensures a more streamlined and efficient handling of VPN connectivity issues, ultimately enhancing the user experience.
+
+## Proposed Solution
+
+We will introduce a new object, the Connection Manager, which will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionManager::State`. The Connection Manager will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful.
+
+In the event of a probe failure, the Connection Manager will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted. Once the issue has been resolved, the cycle of probes commences anew. If there are no ongoing probes, the Connection Manager will enter an Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or `NoSignal` (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Manager.
```

Comment:
I think this doc needs to discuss how ConnectionHealth can also change the UI for users. 

And what if we changed that current practice as part of this work - what if a bad periodic ConnectionHealth check just signals ConnectionManager to run its checks (if it's currently in idle, and not already checking)?

Reply:
> I think this doc needs to discuss how ConnectionHealth can also change the UI for users.

Are you referring to the existing stability modes (stable, nosignal, unstable) in the UI, or something different? 

> And what if we changed that current practice as part of this work - what if a bad periodic ConnectionHealth check just signals ConnectionManager to run its checks (if it's currently in idle, and not already checking)?

I'm hesitant to change this as we didn't talk/discuss this in any of the connection manager discussions. It opens a whole other can of worms which we never agreed on

Reply:
Didn't we discuss that ConnectionManager would rationalize any failures reported by ConnectionHealth (by running the ConnectionManager probes to detect the underlying cause) and report that to the front end. So, ConnectionHealth would not directly report to the front end.

Reply:
If ConnectionManager rationalizes failures reported by ConnectionHealth, it seems extra important to have bad health checks kick off a ConnectionManager run (if it's currently idle).

Reply:
Separately, I don't see the doc discussion limited to in-person discussions, which may be what you're saying (or I may be misunderstanding). I think seeing it written down after a group discussion can allow us to see potential interactions we hadn't previously.

Reply:
> Didn't we discuss that ConnectionManager would rationalize any failures reported by ConnectionHealth (by running the ConnectionManager probes to detect the underlying cause) and report that to the front end. So, ConnectionHealth would not directly report to the front end.

Connection Health will continue to show "No Signal" or "Unstable" to the frontend as it does currently. I

Reply:
> If ConnectionManager rationalizes failures reported by ConnectionHealth, it seems extra important to have bad health checks kick off a ConnectionManager run (if it's currently idle).

@mcleinman I agree 100%. I was holding that off for part 2 but figured I might as well combine the 2 phases in one doc. I actually don't think we should be periodically running the checks while the VPN is active. It only makes sense to do so when the VPN goes into no signal. Updated the tech spec to reflect that

Reply:
> Connection Health will continue to show "No Signal" or "Unstable" to the frontend as it does currently.

If Connection Health detects "No Signal", it appears that there will be two alerts in the client app -- one from Connection Health showing the connection state, and potentially another from Connection Diagnostics indicating the cause of the anomalous connection state. Is that correct? See my other comment about the potential downsides of using a modal.

I think it makes sense to show the connection state alert as soon as possible, because it will take longer to run the Connection Diagnostics probes and determine the cause. If the cause can be determined, the initial alert could be subsequently modified to show both the connection state and cause. For example, the user would see "No Signal"  when Connection Health reports it. It will be replaced by something like "No Signal. Check your internet connection." when Connection Diagnostics completes its probes.

Or two alerts can be shown, one for the connection state and another for the cause.

Consider if Connection Diagnostics should be responsible for emitting all connection state related signals to the font end, for better encapsulation. (Connection Health can live inside Connection Diagnostics and notify it.)

Reply:
After looking at the code, I believe that the health checks do not pop modals - they just end up calling signals that update the color of the icon in the system tray and on the main screen. I may be wrong here, and would love someone else to confirm.

Reply:
I really like the idea of popping a modal as soon as possible, and updating it as we have more information.

Reply:
Thinking more on this - 

If it does work in this way - ConnectionHealth updates colors and text on main screen, the connection diagnostic checks are what ends up popping the modals - that seems fair, but could be unintuitive to future engineers down the line. Can we leave some wayfinding comments in the code (both in ConnectionHealth and in ConnectionDiagnostics) explaining this setup?

Reply:
@vinocher @mcleinman Yes the way I had envisioned this would be that the VPN goes into no signal as it does currently, and the text updates in the controller to show the new stability value (e.g. `Unstable`). Then as we go through the various probes, as soon as the first one fails and we are able to determine the cause for network issues, we will show a popup/modal with the reason. The user has close this modal as they wish and they'll still be able to see the No Signal text in the controller. Does that sound good?